File: rails/app/models/draft_pick.rb
class DraftPick < ApplicationRecord
  self.table_name = 'draft_picks'

  belongs_to :participant, class_name: 'DraftParticipant', inverse_of: :picks
  belongs_to :card, class_name: 'Card', inverse_of: :draft_picks

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:pick_number_positive, 'Pick number must be greater than zero') unless ((pick_number.nil? || pick_number > 0))
    errors.add(:pack_number_range, 'Pack number must be between 1 and 3') unless ((pack_number.nil? || (pack_number >= 1 && pack_number <= 3)))
  end

  def to_s
    pick_number.to_s
  end

  # Business operations

  def is_first_pick
    # TODO: implement is_first_pick
    nil
  end

  def as_json(options = {})
    hash = super(options)
    hash['pickedAt'] = hash.delete('picked_at') if hash.key?('picked_at')
    hash
  end
end

File: phoenix/lib/cards_project/content/draft_pick.ex
defmodule CardsProject.Content.DraftPick do
  use Ecto.Schema
  import Ecto.Changeset

  schema "draft_picks" do
    field :pick_number, :integer
    field :pack_number, :integer
    field :picked_at, :naive_datetime
    belongs_to :participant, CardsProject.Content.DraftParticipant
    belongs_to :card, CardsProject.Cards.Card

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:pick_number, :pack_number, :picked_at, :participant_id, :card_id])
    |> validate_required([:pick_number, :pack_number, :picked_at])
    |> validate_number(:pick_number, greater_than: 0, message: "Pick number must be greater than zero")
    |> validate_number(:pack_number, greater_than_or_equal_to: 1, less_than_or_equal_to: 3, message: "Pack number must be between 1 and 3")
  end

  # ── Business operations ────────────────────────────────────────────

  def is_first_pick(_record) do
    # TODO: implement DraftPick.is_first_pick
    {:error, :not_implemented}
  end
end
